src/vmm/src/lib.rs

```diff
@@ -493,6 +498,116 @@ impl Vmm {
+    fn latest_balloon_stats(&self) -> std::result::Result<VmmData, VmmActionError> {
+        if let Some(busdev) = self.get_bus_device(DeviceType::Virtio(TYPE_BALLOON), BALLOON_DEV_ID)
+        {
+            let virtio_device = busdev
+                .lock()
+                .expect("Poisoned lock")
+                .as_any()
+                .downcast_ref::<MmioTransport>()
+                // Only MmioTransport implements BusDevice at this point.
+                .expect("Unexpected BusDevice type")
+                .device();
+
+            let latest_stats = virtio_device
+                .lock()
+                .expect("Poisoned lock")
+                .as_mut_any()
+                .downcast_mut::<Balloon>()
+                .unwrap()
+                .latest_stats()
+                .map(|stats| VmmData::BalloonStats(stats.clone()));
+
+            latest_stats.ok_or(VmmActionError::BalloonConfig(
+                BalloonConfigError::StatsNotFound,
+            ))
+        } else {
+            Err(VmmActionError::BalloonConfig(
+                BalloonConfigError::DeviceNotFound,
+            ))
+        }
+    }
+
+    /// Updates configuration for the balloon device as described in `balloon_update`.
+    fn update_balloon_config(&mut self, balloon_update: BalloonUpdateConfig) -> ActionResult {
+        if let Some(busdev) = self.get_bus_device(DeviceType::Virtio(TYPE_BALLOON), BALLOON_DEV_ID)
+        {
+            {
+                let virtio_device = busdev
+                    .lock()
+                    .expect("Poisoned lock")
+                    .as_any()
+                    .downcast_ref::<MmioTransport>()
+                    // Only MmioTransport implements BusDevice at this point.
+                    .expect("Unexpected BusDevice type")
+                    .device();
+
+                virtio_device
+                    .lock()
+                    .expect("Poisoned lock")
+                    .as_mut_any()
+                    .downcast_mut::<Balloon>()
+                    .unwrap()
+                    .update_size(balloon_update.amount_mb)
+                    .map_err(|_| {
+                        VmmActionError::BalloonConfig(BalloonConfigError::MemorySizeOverflow)
+                    })?;
+            }
+
+            let locked_dev = busdev.lock().expect("Poisoned lock");
+            locked_dev
+                .interrupt(devices::virtio::VIRTIO_MMIO_INT_CONFIG)
+                .map_err(|err| {
+                    VmmActionError::BalloonConfig(BalloonConfigError::UpdateFailure(err))
+                })?;
+        } else {
+            return Err(VmmActionError::BalloonConfig(
+                BalloonConfigError::DeviceNotFound,
+            ));
+        }
+
+        Ok(())
+    }
+
+    /// Updates configuration for the balloon device as described in `balloon_stats_update`.
+    fn update_balloon_stats_config(
+        &mut self,
+        balloon_stats_update: BalloonUpdateStatsConfig,
```

Comment:
This can directly take `stats_polling_interval_s`.

Either that, or define `BalloonUpdateStatsConfig` as part of the `devices::virtio::balloon` if it makes sense.

Reply:
Why `devices::virtio::balloon` and not `vmm::vmm_config::balloon`?

Also, I would prefer not to completely drop these structures as they are used in deserializing the API requests, but I agree the vmm functions should take the inner directly.

Reply:
I don't have a strong preference where we define the structures as long as we don't break the logical dependency tree. `vmm_config` might be confusingly named, but if you look inside it you can see it defines API and RPC level configuration structures. The API component sits on top of RPC which sits on top of `Vmm`. We want the `Vmm` to be independent of API and RPC therefore we shouldn't take any dependencies like that.

We have an uncompleted decoupling item https://github.com/firecracker-microvm/firecracker/issues/1713 that would actually enforce the above and eliminate confusions.